File: src/precalculus_law_of_sines_cosines_hard.py
"""
precalculus - law_of_sines_cosines (hard)
Generated: 2026-02-22T04:51:09.857760
"""

from problem_utils import *
from svg_utils import Diagram

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Ambiguous case (SSA) with two solutions
        side_a = randint(8, 15)
        angle_A = randint(25, 40)
        side_b = randint(side_a + 2, side_a + 8)
        
        # Calculate angle B using law of sines (will have two solutions)
        sin_B = side_b * sin(angle_A * pi / 180) / side_a
        angle_B1 = float(asin(sin_B) * 180 / pi)
        angle_B2 = 180 - angle_B1
        
        # Check both are valid
        angle_C1 = 180 - angle_A - angle_B1
        angle_C2 = 180 - angle_A - angle_B2
        
        if angle_C1 > 0 and angle_C2 > 0:
            # Find side c for both triangles
            side_c1 = side_a * sin(angle_C1 * pi / 180) / sin(angle_A * pi / 180)
            side_c2 = side_a * sin(angle_C2 * pi / 180) / sin(angle_A * pi / 180)
            
            area1 = 0.5 * side_a * side_b * sin(angle_C1 * pi / 180)
            area2 = 0.5 * side_a * side_b * sin(angle_C2 * pi / 180)
            
            total_area = float(area1 + area2)
            ans = round(total_area, 2)
            
            d = Diagram(width=350, height=200, padding=30)
            p1 = (0, 0)
            p2 = (side_a, 0)
            angle_rad = angle_A * pi / 180
            p3_1 = (side_b * cos(angle_rad), side_b * sin(angle_rad))
            
            d.polygon([p1, p2, p3_1], labels=['B', 'C', 'A'])
            d.segment_label(p1, p2, f'{side_a}')
            d.segment_label(p1, p3_1, f'{side_b}')
            d.angle_arc(p1, p2, p3_1, f'{angle_A}°')
            
            return problem(
                question=f"In triangle $ABC$, $a = {side_a}$, $b = {side_b}$, and $\\angle A = {angle_A}°$. This is an ambiguous case (SSA) that yields two possible triangles. Find the sum of the areas of both possible triangles.",
                answer=ans,
                difficulty=(1700, 1850),
                topic="precalculus/law_of_sines_cosines",
                solution=steps(
                    f"Use Law of Sines: $\\frac{{\\sin B}}{{b}} = \\frac{{\\sin A}}{{a}}$",
                    f"$\\sin B = \\frac{{{side_b} \\cdot \\sin({angle_A}°)}}{{{side_a}}} \\approx {float(sin_B):.4f}$",
                    f"Two solutions: $B_1 \\approx {angle_B1:.2f}°$ and $B_2 \\approx {angle_B2:.2f}°$",
                    f"Triangle 1: $C_1 = {angle_C1:.2f}°$, $c_1 \\approx {float(side_c1):.2f}$",
                    f"Area$_1 = \\frac{{1}}{{2}} \\cdot {side_a} \\cdot {side_b} \\cdot \\sin({angle_C1:.2f}°) \\approx {float(area1):.2f}$",
                    f"Triangle 2: $C_2 = {angle_C2:.2f}°$, $c_2 \\approx {float(side_c2):.2f}$",
                    f"Area$_2 = \\frac{{1}}{{2}} \\cdot {side_a} \\cdot {side_b} \\cdot \\sin({angle_C2:.2f}°) \\approx {float(area2):.2f}$",
                    f"Total area = ${float(area1):.2f} + {float(area2):.2f} = {ans}$"
                ),
                calculator="scientific",
                image=d.render()
            )
    
    if problem_type == 2:
        # Law of cosines application with area calculation
        side_a = randint(10, 20)
        side_b = randint(10, 20)
        angle_C = randint(50, 130)
        
        # Calculate side c using law of cosines
        side_c_sq = side_a**2 + side_b**2 - 2*side_a*side_b*cos(angle_C * pi / 180)
        side_c = sqrt(side_c_sq)
        
        # Calculate area using formula
        area = 0.5 * side_a * side_b * sin(angle_C * pi / 180)
        ans = round(float(area), 2)
        
        d = Diagram(width=300, height=250, padding=30)
        p1 = (0, 0)
        p2 = (side_b, 0)
        angle_rad = angle_C * pi / 180
        p3 = (side_a * cos(angle_rad), side_a * sin(angle_rad))
        
        d.polygon([p1, p2, p3], labels=['C', 'A', 'B'])
        d.segment_label(p1, p3, f'{side_a}')
        d.segment_label(p1, p2, f'{side_b}')
        d.angle_arc(p1, p2, p3, f'{angle_C}°')
        
        return problem(
            question=f"In triangle $ABC$, $a = {side_a}$, $b = {side_b}$, and $\\angle C = {angle_C}°$. Find the area of the triangle.",
            answer=ans,
            difficulty=(1600, 1750),
            topic="precalculus/law_of_sines_cosines",
            solution=steps(
                f"Use the area formula: $\\text{{Area}} = \\frac{{1}}{{2}}ab\\sin C$",
                f"$\\text{{Area}} = \\frac{{1}}{{2}} \\cdot {side_a} \\cdot {side_b} \\cdot \\sin({angle_C}°)$",
                f"$\\text{{Area}} \\approx {ans}$"
            ),
            calculator="scientific",
            image=d.render()
        )
    
    if problem_type == 3:
        # Challenging problem: find angle using law of cosines after using law of sines
        side_a = randint(12, 18)
        angle_A = randint(45, 75)
        angle_B = randint(40, 90 - angle_A)
        angle_C = 180 - angle_A - angle_B
        
        # Calculate side b using law of sines
        side_b = side_a * sin(angle_B * pi / 180) / sin(angle_A * pi / 180)
        
        # Calculate side c using law of sines
        side_c = side_a * sin(angle_C * pi / 180) / sin(angle_A * pi / 180)
        
        # Now find largest angle (opposite to largest side)
        sides = [(side_a, angle_A, 'A'), (float(side_b), angle_B, 'B'), (float(side_c), angle_C, 'C')]
        sides.sort(reverse=True)
        largest_angle = sides[0][1]
        
        ans = round(float(largest_angle), 1)
        
        return problem(
            question=f"In triangle $ABC$, $a = {side_a}$, $\\angle A = {angle_A}°$, and $\\angle B = {angle_B}°$. Find the measure of the largest angle in the triangle.",
            answer=ans,
            difficulty=(1650, 1800),
            topic="precalculus/law_of_sines_cosines",
            solution=steps(
                f"First find $\\angle C = 180° - {angle_A}° - {angle_B}° = {angle_C}°$",
                f"Use Law of Sines to find $b$: $\\frac{{b}}{{\\sin B}} = \\frac{{a}}{{\\sin A}}$",
                f"$b = \\frac{{{side_a} \\cdot \\sin({angle_B}°)}}{{\\sin({angle_A}°)}} \\approx {float(side_b):.2f}$",
                f"Similarly, $c = \\frac{{{side_a} \\cdot \\sin({angle_C}°)}}{{\\sin({angle_A}°)}} \\approx {float(side_c):.2f}$",
                f"The largest angle is opposite the largest side",
                f"Comparing sides: $a = {side_a}$, $b \\approx {float(side_b):.2f}$, $c \\approx {float(side_c):.2f}$",
                f"The largest angle is ${ans}°$"
            ),
            calculator="scientific"
        )
    
    # problem_type == 4
    # Use both laws: law of cosines to find a side, then law of sines to find an angle
    side_a = randint(15, 25)
    side_b = randint(12, 22)
    angle_C = randint(65, 115)
    
    # Find side c using law of cosines
    side_c_sq = side_a**2 + side_b**2 - 2*side_a*side_b*cos(angle_C * pi / 180)
    side_c = sqrt(side_c_sq)
    
    # Find angle A using law of sines
    sin_A = side_a * sin(angle_C * pi / 180) / side_c
    angle_A = asin(sin_A) * 180 / pi
    
    ans = round(float(angle_A), 1)
    
    return problem(
        question=f"In triangle $ABC$, $a = {side_a}$, $b = {side_b}$, and $\\angle C = {angle_C}°$. Find the measure of $\\angle A$.",
        answer=ans,
        difficulty=(1700, 1850),
        topic="precalculus/law_of_sines_cosines",
        solution=steps(
            f"First use Law of Cosines to find side $c$:",
            f"$c^2 = a^2 + b^2 - 2ab\\cos C$",
            f"$c^2 = {side_a}^2 + {side_b}^2 - 2({side_a})({side_b})\\cos({angle_C}°)$",
            f"$c \\approx {float(side_c):.2f}$",
            f"Now use Law of Sines: $\\frac{{\\sin A}}{{a}} = \\frac{{\\sin C}}{{c}}$",
            f"$\\sin A = \\frac{{{side_a} \\cdot \\sin({angle_C}°)}}{{{float(side_c):.2f}}}$",
            f"$\\angle A \\approx {ans}°$"
        ),
        calculator="scientific"
    )

emit(generate())

File: src/precalculus_law_of_sines_cosines_hard.jl
# precalculus - law_of_sines_cosines (hard)
# Generated: 2026-03-08T20:32:26.353702

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script a b c A B C begin
    set_topic!("precalculus/law_of_sines_cosines")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Law of Cosines: given three sides, find largest angle (expert level)
        side_a = randint(8, 25)
        side_b = randint(8, 25)
        side_c = randint(max(abs(side_a - side_b) + 1, 8), min(side_a + side_b - 1, 25))
        
        # Find largest angle (opposite largest side)
        sides = sort([side_a, side_b, side_c], rev=true)
        largest = sides[1]
        other1 = sides[2]
        other2 = sides[3]
        
        # Law of cosines: c² = a² + b² - 2ab cos(C)
        cos_angle = (other1^2 + other2^2 - largest^2) / (2 * other1 * other2)
        angle_rad = acos(cos_angle)
        angle_deg = angle_rad * 180 / π
        
        problem(
            question="A triangle has sides of length $(side_a), $(side_b), and $(side_c). Find the measure of the largest angle in degrees. Round to two decimal places.",
            answer=round(angle_deg, digits=2),
            difficulty=(1850, 2100),
            solution=steps(
                "The largest angle is opposite the largest side (length $(largest))",
                sol("Law of Cosines", "c^2 = a^2 + b^2 - 2ab\\cos(C)"),
                "\\cos(C) = \\frac{$(other1)^2 + $(other2)^2 - $(largest)^2}{2($(other1))($(other2))} = \\frac{$(other1^2 + other2^2 - largest^2)}{$(2*other1*other2)}",
                "C = \\cos^{-1}\\left($(round(cos_angle, digits=4))\\right) \\approx $(round(angle_deg, digits=2))°"
            ),
            calculator="scientific",
            time=180
        )
        
    elseif problem_type == 2
        # Law of Sines with ambiguous case (SSA - competition entry)
        side_a = randint(10, 30)
        angle_A_deg = randint(25, 55)
        angle_A_rad = angle_A_deg * π / 180
        
        # Choose side_b to create ambiguous case
        h = side_a * sin(angle_A_rad)
        side_b = randint(ceil(Int, h) + 2, side_a + 5)
        
        # Two possible angles for B
        sin_B = side_b * sin(angle_A_rad) / side_a
        
        if sin_B < 1 && side_b > side_a * sin(angle_A_rad) && side_b < side_a
            angle_B1_rad = asin(sin_B)
            angle_B1_deg = angle_B1_rad * 180 / π
            angle_B2_deg = 180 - angle_B1_deg
            
            angle_C1_deg = 180 - angle_A_deg - angle_B1_deg
            angle_C2_deg = 180 - angle_A_deg - angle_B2_deg
            
            side_c1 = side_a * sin(angle_C1_deg * π / 180) / sin(angle_A_rad)
            side_c2 = side_a * sin(angle_C2_deg * π / 180) / sin(angle_A_rad)
            
            num_triangles = (angle_C1_deg > 0 && angle_C2_deg > 0) ? 2 : 1
            
            problem(
                question="In triangle ABC, a = $(side_a), b = $(side_b), and A = $(angle_A_deg)°. How many distinct triangles can be formed with these measurements?",
                answer=num_triangles,
                difficulty=(2100, 2400),
                solution=steps(
                    "Check for ambiguous case (SSA): given two sides and non-included angle",
                    sol("Law of Sines", "\\frac{\\sin A}{a} = \\frac{\\sin B}{b}"),
                    "\\sin B = \\frac{b \\sin A}{a} = \\frac{$(side_b) \\sin($(angle_A_deg)°)}{$(side_a)} \\approx $(round(sin_B, digits=4))",
                    "Since 0 < \\sin B < 1, angle B has two possible values: B₁ ≈ $(round(angle_B1_deg, digits=2))° and B₂ ≈ $(round(angle_B2_deg, digits=2))°",
                    "Check if both create valid triangles (C > 0°): $(num_triangles) distinct triangle(s) possible"
                ),
                calculator="scientific",
                time=240
            )
        else
            # Fallback to simpler unique triangle case
            side_b = randint(side_a + 3, side_a + 15)
            sin_B = side_b * sin(angle_A_rad) / side_a
            sin_B = min(sin_B, 0.99)
            
            angle_B_deg = round(asin(sin_B) * 180 / π, digits=2)
            
            problem(
                question="In triangle ABC, a = $(side_a), b = $(side_b), and A = $(angle_A_deg)°. Find angle B in degrees. Round to two decimal places.",
                answer=angle_B_deg,
                difficulty=(1800, 2000),
                solution=steps(
                    sol("Law of Sines", "\\frac{\\sin A}{a} = \\frac{\\sin B}{b}"),
                    "\\sin B = \\frac{b \\sin A}{a} = \\frac{$(side_b) \\sin($(angle_A_deg)°)}{$(side_a)}",
                    "B = \\sin^{-1}\\left($(round(sin_B, digits=4))\\right) \\approx $(angle_B_deg)°"
                ),
                calculator="scientific",
                time=150
            )
        end
        
    elseif problem_type == 3
        # Law of Cosines to find third side (SAS)
        side_a = randint(12, 40)
        side_b = randint(12, 40)
        angle_C_deg = randint(35, 145)
        angle_C_rad = angle_C_deg * π / 180
        
        # c² = a² + b² - 2ab cos(C)
        side_c_sq = side_a^2 + side_b^2 - 2*side_a*side_b*cos(angle_C_rad)
        side_c = sqrt(side_c_sq)
        
        problem(
            question="In triangle ABC, a = $(side_a), b = $(side_b), and C = $(angle_C_deg)°. Find the length of side c. Round to two decimal places.",
            answer=round(side_c, digits=2),
            difficulty=(1850, 2050),
            solution=steps(
                "Use Law of Cosines with SAS (side-angle-side)",
                sol("c^2 = a^2 + b^2 - 2ab\\cos C"),
                "c^2 = $(side_a)^2 + $(side_b)^2 - 2($(side_a))($(side_b))\\cos($(angle_C_deg)°)",
                "c^2 = $(side_a^2) + $(side_b^2) - $(2*side_a*side_b)\\cos($(angle_C_deg)°) \\approx $(round(side_c_sq, digits=2))",
                "c \\approx $(round(side_c, digits=2))"
            ),
            calculator="scientific",
            time=150
        )
        
    elseif problem_type == 4
        # Area using Law of Sines approach (competition level)
        side_a = randint(15, 35)
        angle_B_deg = randint(40, 80)
        angle_C_deg = randint(40, 80)
        
        if angle_B_deg + angle_C_deg >= 180
            angle_C_deg = 180 - angle_B_deg - randint(20, 40)
        end
        
        angle_A_deg = 180 - angle_B_deg - angle_C_deg
        angle_A_rad = angle_A_deg * π / 180
        angle_B_rad = angle_B_deg * π / 180
        angle_C_rad = angle_C_deg * π / 180
        
        # Find other sides using Law of Sines
        side_b = side_a * sin(angle_B_rad) / sin(angle_A_rad)
        
        # Area = (1/2)ab sin(C)
        area = 0.5 * side_a * side_b * sin(angle_C_rad)
        
        problem(
            question="In triangle ABC, a = $(side_a), B = $(angle_B_deg)°, and C = $(angle_C_deg)°. Find the area of the triangle. Round to two decimal places.",
            answer=round(area, digits=2),
            difficulty=(2000, 2300),
            solution=steps(
                "First find angle A: A = 180° - $(angle_B_deg)° - $(angle_C_deg)° = $(angle_A_deg)°",
                sol("Law of Sines", "\\frac{a}{\\sin A} = \\frac{b}{\\sin B}"),
                "b = \\frac{a \\sin B}{\\sin A} = \\frac{$(side_a) \\sin($(angle_B_deg)°)}{\\sin($(angle_A_deg)°)} \\approx $(round(side_b, digits=2))",
                sol("Area", "\\text{Area} = \\frac{1}{2}ab\\sin C"),
                "\\text{Area} = \\frac{1}{2}($(side_a))($(round(side_b, digits=2)))\\sin($(angle_C_deg)°) \\approx $(round(area, digits=2))"
            ),
            calculator="scientific",
            time=210
        )
        
    elseif problem_type == 5
        # Combined Law of Cosines and Sines (expert)
        side_a = randint(18, 35)
        side_b = randint(18, 35)
        side_c = randint(max(abs(side_a - side_b) + 2, 15), min(side_a + side_b - 2, 40))
        
        # Find angle A using Law of Cosines
        cos_A = (side_b^2 + side_c^2 - side_a^2) / (2 * side_b * side_c)
        angle_A_rad = acos(cos_A)
        
        # Area using Heron's formula or (1/2)bc sin(A)
        area = 0.5 * side_b * side_c * sin(angle_A_rad)
        
        problem(
            question="A triangle has sides a = $(side_a), b = $(side_b), and c = $(side_c). Find the area of the triangle. Round to two decimal places.",
            answer=round(area, digits=2),
            difficulty=(2100, 2400),
            solution=steps(
                "Use Law of Cosines to find an angle first",
                "\\cos A = \\frac{b^2 + c^2 - a^2}{2bc} = \\frac{$(side_b^2) + $(side_c^2) - $(side_a^2)}{2($(side_b))($(side_c))} = \\frac{$(side_b^2 + side_c^2 - side_a^2)}{$(2*side_b*side_c)}",
                "A = \\cos^{-1}\\left($(round(cos_A, digits=4))\\right)",
                sol("Area", "\\text{Area} = \\frac{1}{2}bc\\sin A"),
                "\\text{Area} = \\frac{1}{2}($(side_b))($(side_c))\\sin A \\approx $(round(area, digits=2))"
            ),
            calculator="scientific",
            time=240
        )
        
    else
        # Distance between two points using Law of Cosines (application problem)
        dist1 = randint(8, 25)
        dist2 = randint(8, 25)
        angle_deg = randint(30, 150)
        angle_rad = angle_deg * π / 180
        
        # Distance = sqrt(d1² + d2² - 2*d1*d2*cos(angle))
        distance_sq = dist1^2 + dist2^2 - 2*dist1*dist2*cos(angle_rad)
        distance = sqrt(distance_sq)
        
        problem(
            question="Two ships leave a port at the same time. One travels $(dist1) km on a bearing, and the other travels $(dist2) km on a bearing that differs by $(angle_deg)° from the first. How far apart are the two ships? Round to two decimal places.",
            answer=round(distance, digits=2),
            difficulty=(1900, 2200),
            solution=steps(
                "The angle between the two paths is $(angle_deg)°",
                sol("Law of Cosines", "d^2 = a^2 + b^2 - 2ab\\cos\\theta"),
                "d^2 = $(dist1)^2 + $(dist2)^2 - 2($(dist1))($(dist2))\\cos($(angle_deg)°)",
                "d^2 = $(dist1^2) + $(dist2^2) - $(2*dist1*dist2)\\cos($(angle_deg)°) \\approx $(round(distance_sq, digits=2))",
                "d \\approx $(round(distance, digits=2)) \\text{ km}"
            ),
            calculator="scientific",
            time=180
        )
    end
end